### PR TITLE
Full JMAP support for forwarded messages

### PR DESCRIPTION
**Pull request by chibenwa**
This is my proposal for this feature.

Hence: 

 - https://github.com/linagora/jmap/commit/3d60d48552320de78db535988d451023359625a4 adds a isFowarded property to the messages in a compatible way with Thunderbird, Roundcube and most MUA using the defacto standard $Fowarded user flag
 - https://github.com/linagora/jmap/commit/c4341eadea92b198a371e9e9afd876c159f33abb adds the opportunity to search this property
 - https://github.com/linagora/jmap/commit/53f36b395d32b38e465d379a425f3a267b04d9ef adds a mechanism for automatically marking a message as forwarded when sending it
 - https://github.com/linagora/jmap/commit/6b9638fc197b9e3d20f97600699f73fc7fd120ea displays a forwarded draft message under the given message and its answers
 - https://github.com/linagora/jmap/commit/7b7f0d7fe17f46f5800659758c8383b1c7c73b7a updates the thread algorithm to take into account forwarded messages.

We need to push the last changes on master to get a more "readable" PR. Could someone give me the rights?

### CODE REVIEW

File: spec/mail/message.mdown [hunk 1]
```diff
@@ -17,6 +17,8 @@ A **Message** object has the following properties:
   The ids of the mailboxes the message is in. A message MUST belong to one or more mailboxes at all times (until it is deleted).
 - **inReplyToMessageId**: `String|null`
   The id of the Message this message is a reply to. This is primarily for drafts, but the server MAY support this for received messages as well by looking up the [@!RFC5322] Message-Id referenced in the `In-Reply-To` header and searching for this message in the user's mail.
+- **fowardingMessageId**: `String|null`
+  The id of the Message this message is fowarding. This is primarily for drafts, but the server MAY support this for received messages as well by looking up the [@!RFC5322] Message-Id referenced in the `In-Reply-To` header and searching for this message in the user's mail.
```
Comment (rouazana) [inline spec/mail/message.mdown]: Maybe you meant `X-Forwarded-Message-Id`?

File: spec/mail/message.mdown [hunk 2]
```diff
@@ -259,6 +261,8 @@ The properties of the Message object submitted for creation MUST conform to the
 - **threadId**: This property MUST NOT be included. It is set by the server upon creation.
 - **mailboxIds**: This property MUST be included. The value MUST include the id of either the mailbox with `role == "drafts"` (to save a draft) or the mailbox with `role == "outbox"` (to send the message). If this mailbox does not have `mustBeOnlyMailbox == true`, others may be included too.
 - **inReplyToMessageId**: Optional. If included, the server will look up this message and if found set appropriate `References` and `In-Reply-To` headers. These will override any such headers supplied in the *headers* property. If not found, the creation MUST be rejected with an `inReplyToNotFound` error.
+- **fowardingMessageId**: Optional. If included, the server will look up this message and if found set appropriate `References` and `X-Forwarded-Message-Id` headers. These will override any such headers 
+supplied in the *headers* property. If not found, the creation MUST be rejected with an `fowardingNotFound` error.
```
Comment (rouazana) [inline spec/mail/message.mdown]: s/an `fowardingNotFound`/a `fowardingNotFound`/

Comment (rouazana) [reply spec/mail/message.mdown]: I just tested with RoundCube, but it seems it fills only In-Reply-To and References headers. Do you know if the X-Forwarded-Message-Id header is well spread?

File: spec/mail/message.mdown [hunk 3]
```diff
@@ -317,7 +321,11 @@ To send a message, either create a new message directly into the mailbox with `r
...
+When the message is sent, the server MUST delete the message from the **outbox** and SHOULD create a **new** copy of the sent message (with a new id) in the **sent** mailbox, unless the user has indicated another preference. 
+
+If `inReplyToMessageId` was set, the server SHOULD mark this message as `isAnswered: true` at this point, if found. The server is responsible for either reporting an error (normally a "bounce" email), or ensuring delivery of the message to the next hop.
+
+If `fowardingMessageId` was set, the server SHOULD mark this message as `isFowarded: true` at this point, if found. The server is responsible for either reporting an error (normally a "bounce" email), or ensuring delivery of the message to the next hop.
```
Comment (rouazana) [inline spec/mail/message.mdown]: > The server is responsible for either reporting an error (normally a "bounce" email), or ensuring delivery of the message to the next hop.

This part should not be duplicated, no?

File: server-guide/jmap-server-guide.mdown [hunk 4]
```diff
@@ -56,7 +56,7 @@ Data sync properties:
   (after stripping fwd/re etc.). This is used for checking whether we need to
   split out a new thread when a message arrives as the subject has changed (see data structure 7).
 - **messages**: `ThreadMessage[]`
-  The list of messages belonging to the thread, sorted with drafts immediately after the message they are in reply to, then by date order.
+  The list of messages belonging to the thread, sorted with drafts immediately after the message they are in reply to or then the message they are fowarding, then by date order.
```
Comment (rouazana) [inline server-guide/jmap-server-guide.mdown]: s/or then the/or the/
or better s/or then the message they are/or they are forwarding

File: spec/mail/thread.mdown [hunk 5]
```diff
@@ -14,6 +14,7 @@ A **Thread** object has the following properties:
   The ids of the messages in the thread, sorted such that:
 
     - Any message with `isDraft == true` and an *inReplyToMessageId* property that corresponds to another message in the thread comes immediately after that message in the sort order.
+    - Directly after, any message with `isDraft == true` and an *fowardingMessageId* property that corresponds to another message in the thread comes immediately after that message in the sort order.
```
Comment (rouazana) [inline spec/mail/thread.mdown]: On GMail for example forwarded messages are mixed with replied messages, all being sort by date.
Seems better, no?

File: spec/mail/message.mdown (hunk 2 above)
Comment (chibenwa) [reply spec/mail/message.mdown]: Ok for typo.

It is use at least with thunderbird. I don't know if it is well spread appart this, but we can require the se of this header to make the feature work, no?

File: spec/mail/message.mdown (hunk 3 above)
Comment (chibenwa) [reply spec/mail/message.mdown]: Yes it seems. Thanks

File: spec/mail/thread.mdown (hunk 5 above)
Comment (chibenwa) [reply spec/mail/thread.mdown]: Ok

**Comment by chibenwa** [comment]
Issue opened https://github.com/linagora/jmap/issues/5

File: spec/mail/thread.mdown [hunk 6]
```diff
@@ -13,8 +13,7 @@ A **Thread** object has the following properties:
 - **messageIds**: `String[]`
   The ids of the messages in the thread, sorted such that:
 
-    - Any message with `isDraft == true` and an *inReplyToMessageId* property that corresponds to another message in the thread comes immediately after that message in the sort order.
-    - Directly after, any message with `isDraft == true` and an *fowardingMessageId* property that corresponds to another message in the thread comes immediately after that message in the sort order.
+    - Any message with `isDraft == true` and an *inReplyToMessageId* or **fowardingMessagId** property that corresponds to another message in the thread comes immediately after that message in the sort order.
```
Comment (rouazana) [inline spec/mail/thread.mdown]: why double stars?

File: spec/mail/message.mdown (hunk 2 above)
Comment (rouazana) [reply spec/mail/message.mdown]: Just saw this one: fo***r***wardingNotFound
Else ok to use this header by default.

File: spec/mail/message.mdown (hunk 1 above)
Comment (rouazana) [reply spec/mail/message.mdown]: what about this one?

File: spec/mail/thread.mdown [hunk 7]
```diff
@@ -13,7 +13,7 @@ A **Thread** object has the following properties:
 - **messageIds**: `String[]`
   The ids of the messages in the thread, sorted such that:
 
-    - Any message with `isDraft == true` and an *inReplyToMessageId* or **fowardingMessagId** property that corresponds to another message in the thread comes immediately after that message in the sort order.
+    - Any message with `isDraft == true` and an *inReplyToMessageId* or *fowardingMessagId* property that corresponds to another message in the thread comes immediately after that message in the sort order.
```
Comment (rouazana) [inline spec/mail/thread.mdown]: s/fowardingMessagId/forwardingMessagId/

**Review by rouazana — COMMENTED** [review]
Can you please rebase? The all diff is not readable and I'm not sure all is fine.

**Comment by chibenwa** [comment]
I already rebased on master. All is fine on my laptop.

**Pull request closed without merging**